src/mlpack/methods/lmnn/constraints_impl.hpp

```diff
@@ -0,0 +1,328 @@
+  Precalculate(labels);
+
+  // KNN instance.
+  KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  for (size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Perform KNN search with same class points as both reference
+    // set and query set.
+    knn.Train(dataset.cols(indexSame[i]));
+    knn.Search(k, neighbors, distances);
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexSame[i].at(neighbors(j));
+
+    // Store target neihbors.
+    outputMatrix.cols(indexSame[i]) = neighbors;
+  }
+}
+
+// Calculates k similar labeled nearest neighbors  on a
+// batch of data points.
+template<typename MetricType>
+void Constraints<MetricType>::TargetNeighbors(arma::Mat<size_t>& outputMatrix,
+                                              const arma::mat& dataset,
+                                              const arma::Row<size_t>& labels,
+                                              const size_t begin,
+                                              const size_t batchSize)
+{
+  // Perform pre-calculation. If neccesary.
+  Precalculate(labels);
+
+  arma::mat subDataset = dataset.cols(begin, begin + batchSize - 1);
+  arma::Row<size_t> sublabels = labels.cols(begin, begin + batchSize - 1);
+
+  // KNN instance.
+  KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  // Vectors to store indices.
+  arma::uvec subIndexSame;
+
+  for (size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Calculate Target Neighbors.
+    subIndexSame = arma::find(sublabels == uniqueLabels[i]);
+
+    // Perform KNN search with same class points as both reference
+    // set and query set.
+    knn.Train(dataset.cols(indexSame[i]));
+    knn.Search(subDataset.cols(subIndexSame), k, neighbors, distances);
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexSame[i].at(neighbors(j));
+
+    // Store target neighbors.
+    outputMatrix.cols(begin + subIndexSame) = neighbors;
+  }
+}
+
+// Calculates k differently labeled nearest neighbors.
+template<typename MetricType>
+void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputMatrix,
+                                        const arma::mat& dataset,
+                                        const arma::Row<size_t>& labels)
+{
+  // Perform pre-calculation. If neccesary.
+  Precalculate(labels);
+
+  // KNN instance.
+  KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  for (size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Perform KNN search with differently labeled points as reference
+    // set and  same class points as query set.
+    knn.Train(dataset.cols(indexDiff[i]));
+    knn.Search(dataset.cols(indexSame[i]), k, neighbors, distances);
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexDiff[i].at(neighbors(j));
```

Comment:
Just an idea, but could directly use ``outputMatrix`` and avoid the extra copy at the end?

Reply:
I think we can do that, but will be a bit messy.
```
// Re-map neighbors to their index.
for (size_t j = 0; j < neighbors.n_elem; j++)
  outputMatrix(j % k, indexSame[i](j / k)) = indexSame[i].at(neighbors(j));
```
Shall we do this?

Reply:
Your call, not sure we would see any performance improvements.